snuba/query/processors/timeseries_processor.py

```diff
@@ -185,3 +185,58 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
         condition = query.get_condition_from_ast()
         if condition:
             query.set_ast_condition(condition.transform(self.__process_condition))
+
+
+GRANULARITY_MAPPING = {
+    "toStartOfMinute": 60,
+    "toStartOfHour": 3600,
+    "toDate": 86400,
+}
+
+
+def extract_granularity_from_query(query: Query, column: str) -> Optional[int]:
+    """
+    This extracts the `granularity` from the `groupby` statement of the query.
```

Comment:
The query has a `get_granularity()` function. Why do you need to parse it out of the conditions?

Reply:
This is to be used by code past the time series processor, where the granularity field has already be consumed and it is not guaranteed that what is in the query still reflects it. 
See details here: https://github.com/getsentry/snuba/pull/1608#discussion_r562946915

Specifically this method is used in a storage selector, after the whole logical processing. 
What we could do to simplify this, is, when we will have moved the logical processing to the sentry side, we will be able to rely on the `granularity` field, but we will have to figure out a way to enforce that query processing cannot change the query in a way that makes granularity meaningless.

Reply:
I think I see your point, but why would the granularity field not be valid? If we are running queries that are not the same granularity specified in the original query, isn't that a bug?

Reply:
Likely it would be a bug, with the query processors we have today. You are right on this. Still we pretty much always try to stick to what is in the query clauses as source of truth to keep the processing robust to refactoring and changes in the processors. 
With granularity, I think, the issue is in our query processing that has two sources of truth that are not guaranteed to be in sync. Specifically these are the issues:
- With the old query language granularity should not even be a field in the query class. IT was processed in the extension. There was no reason for it to keep existing.
- In alternative we could have preserved as a top level element in the query but we would need a way to ensure the time column would be resolved later. We never thought through this deeper.
- In SnQL the granularity clause is not mandatory. Which means you can totally produce a valid bucketed query without a granularity field by constructing your bucketed time conditions. So relying on granularity would not fly here.
- You can always provide the granularity and not reference the bucketed column but still GROUP BY `toStartOfHour(timestamp)`. That would be still a valid query to send to the hourly column, but it would not have the granularity field.

I think we need to rethink the way bucketed query are supported in the language to make this consistent and at the same time catch as many query as possible that should go to the hourly column. 
Thoughts ?

Reply:
The "granularity" idea sounds like another candidate to move out of Snuba entirely and have the client provide the right groupby in SnQL. I see your point about SnQL allowing both options, so this parsing remains necessary to handle that case. I can't think offhand of a better way to do this check right now.